webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultCsvValidationService.java

```diff
@@ -0,0 +1,200 @@
+            if (!validationFailures.isEmpty()) continue;
+
+            //Validate the School
+            if (currentSchool == null || !currentSchool.equals(school)) {
+                if (currentSchool != null) completedSchools.add(currentSchool);
+                completedSchoolGroupYears.clear();
+                currentSchoolGroupYear = null;
+                currentSchool = school;
+
+                validateSchool(validationFailures, availableSchools, completedSchools, school, record.getRecordNumber());
+            }
+
+            //Validate the School, Group, and Year
+            final String schoolGroupYear = school + group + String.valueOf(schoolYear);
+            if (currentSchoolGroupYear == null || !currentSchoolGroupYear.equalsIgnoreCase(schoolGroupYear)) {
+                if (currentSchoolGroupYear != null) completedSchoolGroupYears.add(currentSchoolGroupYear);
+                currentSchoolGroupYearSubject = null;
+                currentSchoolGroupYear = schoolGroupYear;
+
+                validateSchoolGroupYear(validationFailures, completedSchoolGroupYears, schoolGroupYear, record.getRecordNumber());
+            }
+
+            //Validate the Subject
+            if (currentSchoolGroupYearSubject == null) {
+                currentSchoolGroupYearSubject = subject;
+            } else if (subject != null && !subject.equals(currentSchoolGroupYearSubject)) {
+                validationFailures.add(CsvValidationFailure.builder()
+                        .row(record.getRecordNumber())
+                        .message("Record is attempting to re-define group subject from [" + currentSchoolGroupYearSubject.name() + "] to [" + subject.name() + "]")
+                        .build());
+                currentSchoolGroupYearSubject = subject;
+            }
+        }
+
+        return validationFailures;
+    }
+
+    private void validateSchoolGroupYear(final List<CsvValidationFailure> validationFailures,
+                                         final Set<String> completedSchoolGroupYears,
+                                         final String schoolGroupYear,
+                                         final long recordNumber) {
+        if (completedSchoolGroupYears.contains(schoolGroupYear)) {
+            validationFailures.add(CsvValidationFailure.builder()
+                    .row(recordNumber)
+                    .message("Record is out-of-order: sort CSV by school, group, year")
+                    .build());
+        }
+    }
+
+    @Override
+    public String toMessage(final List<CsvValidationFailure> failures) {
+        final StringBuilder messageBuilder = new StringBuilder();
+        int maxMessages = MaxFailureMessages;
+        for (final CsvValidationFailure failure : failures) {
+            messageBuilder.append(String.format(ValidationFormat, failure.getRow(), failure.getMessage()));
+            if (maxMessages-- < 0) {
```

Comment:
This whole thing is a bit confusing; it seems like a subList could be used instead. Meh.